design/src/rfcs/rfc0004_retry_behavior.md

```diff
@@ -0,0 +1,147 @@
+
+Terminology
+-----------
+
+- **Shared Config**: An `aws_config::Config` struct that is responsible for storing shared configuration data that is used across all services. This is not generated and lives in the `aws-config` crate.
+- **Service-specific Config**: A code-generated `Config` that has methods for setting service-specific configuration. Each `Config` is defined in the `config` module of its parent service. For example, the S3-specific config struct is `use`able from `aws_sdk_s3::config::Config` and re-exported as `aws_sdk_s3::Config`.
+
+Retry config
+--------------------
+
+Users can set the number of retries in several ways:
+- By setting the `AWS_MAX_ATTEMPTS` environment variable
+- By calling the `aws_config::ConfigLoader::retry_config(..)` method
+- By calling the retry_config method on a service-specific config
+
+The above list is in order of increasing precedence e.g. setting retry attempts with the `retry_config` method will override a value set by `AWS_MAX_ATTEMPTS`.
+
+## Setting an environment variable
+
+Here's an example app that logs your AWS user's identity
+
+```rust
+use aws_sdk_sts as sts;
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = aws_config::load_from_env().await;
+
+    let sts = sts::Client::new(&config);
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+Then, in your terminal:
+
+```sh
+# Set the env var before running the example program
+export AWS_MAX_ATTEMPTS=5
+# Run the example program
+cargo run
+```
+
+## Calling a method on an AWS shared config
+
+Here's an example app that creates a shared config with custom retry behavior and then logs your AWS user's identity
+
+```rust
+use aws_sdk_sts as sts;
+use aws_types::config::Config;
+use aws_types::retry_config::RetryConfig;
+
+async fn create_config() -> aws_types::config::Config {
+    let retry_config = RetryConfig::builder().max_attempts(5).build();
+    let config = aws_config::from_env()
+        .retry_config(retry_config)
+        .load().await;
+
+    config
+}
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = create_config().await;
+
+    let sts = sts::Client::new(&config);
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+## Calling a method on service-specific config
+
+Here's an example app that creates a service-specific config with custom retry behavior and then logs your AWS user's identity
+
+```rust
+use aws_sdk_sts as sts;
+use aws_types::config::Config;
+use aws_types::retry_config::RetryConfig;
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = aws_config::load_from_env().await;
+    let retry_config = RetryConfig::builder().max_attempts(5).build();
```

Comment:
Do you anticipate that `max_attempts` will be the only config option for a long time? For presigning, we added a top-level `expires_in` function to `PresigningConfig` that allowed for bypassing the builder since it only had one field, but still kept the builder so that more fields could be added in the future if needed.

Reply:
How long is a long time? I think it'll be several months at least before we expand it. I think using a builder is definitely more verbose than just adding a field to config structs but I don't think it's any more verbose than how most hings in the SDK are defined.

I think this is a good opportunity to reify our philosophy on how we break up categories of common configuration that get re-used in multiple kinds of config structs. I think that you're right in this case and my only concern is whether or not it confuses developers if we have multiple ways of doing the same thing.